tests/model_inheritance/test_abstract_inheritance.py

```diff
@@ -1,63 +1,106 @@
+        }
+        if field_len is not None:
+            attrs['foo'] = models.CharField(max_length=field_len)
+        if base_names is None:
+            bases = (models.Model,)
+        else:
+            bases = tuple(self.classes[base_name] for base_name in base_names)
+        self.classes[model_name] = type(model_name, bases, attrs)
+
+    def assertCorrectInheritance(self, model_name, field_len):
+        self.assertEqual(self.classes[model_name]._meta.get_field('foo').max_length, field_len)
 
-        class DerivedChild(AbstractBase):
-            name = models.CharField(max_length=50)
-
-        class DerivedGrandChild(AbstractDescendant):
-            pass
-
-        self.assertEqual(AbstractDescendant._meta.get_field('name').max_length, 50)
-        self.assertEqual(DerivedChild._meta.get_field('name').max_length, 50)
-        self.assertEqual(DerivedGrandChild._meta.get_field('name').max_length, 50)
-
-    def test_multiple_inheritance_cannot_shadow_inherited_field(self):
-        class ParentA(models.Model):
-            name = models.CharField(max_length=255)
-
-            class Meta:
-                abstract = True
-
-        class ParentB(models.Model):
-            name = models.IntegerField()
-
-            class Meta:
-                abstract = True
-
-        class Child(ParentA, ParentB):
-            pass
-
-        self.assertEqual(Child.check(), [
-            Error(
-                "The field 'name' clashes with the field 'name' from model "
-                "'model_inheritance.child'.",
-                obj=Child._meta.get_field('name'),
-                id='models.E006',
-            ),
-        ])
+    def test_single_parent(self):
+        """ Test multiple levels of ancestry, but with a single parent at each level.
+            NO = non-overriding, O = overriding.
+        """
+        self._build_class('AbstractBase', None, 30, abstract=True)
+        self._build_class('AbstractDescendantNO', ['AbstractBase'], abstract=True)
+        self._build_class('AbstractDescendantO', ['AbstractBase'], 40, abstract=True)
+        self._build_class('DerivedFromBaseNO', ['AbstractBase'])
+        self._build_class('DerivedFromBaseO', ['AbstractBase'], 50)
+        self._build_class('DerivedFromDescendantNONO', ['AbstractDescendantNO'])
+        self._build_class('DerivedFromDescendantNOO', ['AbstractDescendantNO'], 50)
+        self._build_class('DerivedFromDescendantONO', ['AbstractDescendantO'])
+        self._build_class('DerivedFromDescendantOO', ['AbstractDescendantO'], 50)
+
+        self.assertCorrectInheritance('DerivedFromBaseNO', 30)
+        self.assertCorrectInheritance('DerivedFromBaseO', 50)
+        self.assertCorrectInheritance('DerivedFromDescendantNONO', 30)
+        self.assertCorrectInheritance('DerivedFromDescendantNOO', 50)
+        self.assertCorrectInheritance('DerivedFromDescendantONO', 40)
+        self.assertCorrectInheritance('DerivedFromDescendantOO', 50)
```

Comment:
Appreciate what you've done with the tests but I think they're harder to comprehend now. As a Django dev I can jump in and read model classes, but the helpers are obscuring things a bit to me.

Also you now have multiple tests per test method now - ideally there should just be a handful of classes per test and some assertions. I guess you can split based on overriding, non-overriding, etc.

Reply:
Thanks for taking a look!

Yeah it is more obscure, I was just hoping to find a way to avoid the verbosity of
```

Meta:
    abstract = True
```
over and over again.

Happy to revert, I guess the exercise helped list out several additional cases that weren't being covered before, like making sure each level pulls through from the previous in the absence of overrides.
I agree that test methods should hold minimal assertions that tie more directly to one specific use case, but I found it hard to break these down into multiple methods without having even more repetition/verbosity. Splitting on overriding/non-overriding still fans out a fair amount (if we care for each branch of the fan).

Reply:
Also, question: Is there a pattern for any kind of "short-form" class creation in the django test suite currently? I had a quick look but couldn't find anything, probs due to the "readability" argument.
Could be a worthwhile exercise to implement a standard for "quickbuild" setup functions for model classes -- would be more understandable if such things were used across the board, and much easier to scan than tonnes of vertical scrolling.

Reply:
I'm pretty sure you can make the abstract bases once in the test case class level:

```
class ...Tests(...):
    class SomeBase(models.Model):
         ...
```

Then reuse them appropriately in the individual test methods, which should only need to create the "bottom" classes?

I grepped for `\btype\(` and found a few uses, but none are for "quick" building.

I personally think the "tonnes of vertical scrolling" is a small concern compared to having to grok how `type()` works.

Reply:
> I'm pretty sure you can make the abstract bases once in the test case class level:
> 
> ```
> class ...Tests(...):
>     class SomeBase(models.Model):
>          ...
> ```
> 
> Then reuse them appropriately in the individual test methods, which should only need to create the "bottom" classes?

Ok, I like that. Wasn't aware you could declare there without having isolation or referencing issues, but had a quick play and looks pretty solid.

> I grepped for `\btype\(` and found a few uses, but none are for "quick" building.
> 
> I personally think the "tonnes of vertical scrolling" is a small concern compared to having to grok how `type()` works.

Well the tradeoff is you learn the extra bit of lingo (don't necessarily have to read up on `type`; docstring on util function gives arguably sufficient blackbox-level insight) to then be able to sweep through a whole bunch of class definitions quickly. Having them defined so close together allows for easy comparison between them, as well as between their usage across tests.
I think if someone were to visit this file in the future to work on a single test, then having everything splayed out as normal django class definitions is much more accessible. However, when looking at the tests holistically (to identify all the cases they're covering and how), a compact perspective can be useful.
I've seen a lot of django tests where multiple contributors have added test cases to the same suite, with approaches that are all quite different, with high potential for covering the same functionality multiple times. Seeing more neighbouring tests on the screen when writing one's own can lead to more consistency, both in test layout, and in ensuring the tests all work together to provide maximal non-overlapping coverage.
So I think there's at least _some_ merit to it. But I understand trying to forge a new pattern on a random PR probs serves only to detract from what the PR's discussion should actually be about.

Thanks for having a look anyway, I'll expand it out again after Carlton drops his thoughts on some of the other stuff I mentioned in my main comment (blocking diamond inheritance altogether).